### PR TITLE
Add drain-runners and resume-runners recipes to justfile

### PR DESCRIPTION
**Pull request by jeanschmidt**
**Impact:** OSDC operators performing cluster-wide infrastructure migrations
**Risk:** low

## What
Adds two new `just` recipes (`drain-runners` and `resume-runners`) and a supporting Python helper (`runner_max_map.py`) to safely quiesce and restore the ARC runner fleet during maintenance windows.

## Why
The upcoming IPv4 capacity expansion (`INCREASE_IPV4.md`) requires a drain-apply-resume cutover cycle per cluster — the entire PR stack (VPC secondary CIDRs, Custom Networking, prefix delegation, NAT topology refactor) deploys atomically while the runner fleet is drained. These recipes automate steps 2a and the out-of-band recovery path described in that plan.

## How
- `drain-runners` patches every `AutoscalingRunnerSet` to `maxRunners=0`, taints runner nodes via `just taint-nodes`, then polls for in-flight runner pods to finish (default 1h timeout, configurable via `OSDC_DRAIN_TIMEOUT_SECS`). On timeout, it lists stragglers and provides operator guidance (wait or force-terminate).
- `resume-runners` rebuilds the desired `{ars_name: maxRunners}` map from runner def files and patches each live `AutoscalingRunnerSet` back to its configured value. It also untaints runner nodes. This is kept as out-of-band recovery — the standard cutover path restores runners via `just deploy`.
- `runner_max_map.py` computes the ARS name-to-maxRunners mapping by reading `clusters.yaml` and `modules/arc-runners*/defs/*.yaml`. It imports `resolve_value` from `generate_runners.py` to match the exact whole-block config resolution the chart uses for ARS naming — preventing name mismatches that would cause silent resume failures.

## Changes
- **`osdc/justfile`**: Added `drain-runners` and `resume-runners` recipes
- **`osdc/modules/arc-runners/scripts/python/runner_max_map.py`**: New helper that computes `{AutoscalingRunnerSet name: maxRunners}` from def files for a given cluster, matching the chart's naming transform (underscore-to-dash, prefix resolution)
- **`osdc/modules/arc-runners/scripts/python/test_runner_max_map.py`**: Comprehensive test suite (694 lines) covering all public functions, edge cases, config resolution divergence scenarios, and end-to-end CLI behavior

## Notes
- `resume-runners` is intentionally not part of the standard cutover flow — `just deploy` re-syncs `maxRunners` from defs as part of the normal deploy pipeline. `resume-runners` exists for out-of-band recovery if the deploy path is blocked.
- The drain timeout defaults to 1 hour (covers typical PyTorch suite duration). Override with `OSDC_DRAIN_TIMEOUT_SECS=<seconds>` for longer windows.
- Force-terminating straggler pods aborts the GitHub Actions job mid-execution and may leave dangling runner registrations — coordinate with pytorch/pytorch on-call before using.

## Testing
```
 $  just drain-runners arc-staging
Updating kubeconfig for pytorch-arc-staging (us-west-1)...
Updated context pytorch-arc-staging in /Users/jschmidt/.kube/config
Draining ARC runner scale sets for arc-staging (pytorch-arc-staging)...

  c-mt-l-arm64g2-6-32 — drained (maxRunners=0)
  c-mt-l-arm64g3-16-62 — drained (maxRunners=0)
  c-mt-l-arm64g3-61-463 — drained (maxRunners=0)
  c-mt-l-arm64g4-16-62 — drained (maxRunners=0)
  c-mt-l-barm64g4-62-226 — drained (maxRunners=0)
  c-mt-l-bx86iamx-92-167 — drained (maxRunners=0)
  c-mt-l-bx86iavx512-88-1000-a100-8 — drained (maxRunners=0)
  c-mt-l-bx86iavx512-94-344-t4-8 — drained (maxRunners=0)
  c-mt-l-x86aavx2-189-704-a10g-8 — drained (maxRunners=0)
  c-mt-l-x86aavx2-29-113-a10g — drained (maxRunners=0)
  c-mt-l-x86aavx2-29-113-l4 — drained (maxRunners=0)
  c-mt-l-x86aavx2-45-167-a10g-4 — drained (maxRunners=0)
  c-mt-l-x86aavx2-45-172-l4-4 — drained (maxRunners=0)
  c-mt-l-x86aavx512-125-463 — drained (maxRunners=0)
  c-mt-l-x86iamx-14-27 — drained (maxRunners=0)
  c-mt-l-x86iamx-16-128 — drained (maxRunners=0)
  c-mt-l-x86iamx-22-41 — drained (maxRunners=0)
  c-mt-l-x86iamx-32-128 — drained (maxRunners=0)
  c-mt-l-x86iamx-46-84 — drained (maxRunners=0)
  c-mt-l-x86iamx-8-16 — drained (maxRunners=0)
  c-mt-l-x86iamx-8-32 — drained (maxRunners=0)
  c-mt-l-x86iamx-8-64 — drained (maxRunners=0)
  c-mt-l-x86iavx2-40-160 — drained (maxRunners=0)
  c-mt-l-x86iavx2-8-32 — drained (maxRunners=0)
  c-mt-l-x86iavx512-11-125-a100 — drained (maxRunners=0)
  c-mt-l-x86iavx512-16-128 — drained (maxRunners=0)
  c-mt-l-x86iavx512-16-32 — drained (maxRunners=0)
  c-mt-l-x86iavx512-2-4 — drained (maxRunners=0)
  c-mt-l-x86iavx512-22-250-a100-2 — drained (maxRunners=0)
  c-mt-l-x86iavx512-29-115-t4 — drained (maxRunners=0)
  c-mt-l-x86iavx512-32-256 — drained (maxRunners=0)
  c-mt-l-x86iavx512-37-68 — drained (maxRunners=0)
  c-mt-l-x86iavx512-44-500-a100-4 — drained (maxRunners=0)
  c-mt-l-x86iavx512-45-172-t4-4 — drained (maxRunners=0)
  c-mt-l-x86iavx512-46-85 — drained (maxRunners=0)
  c-mt-l-x86iavx512-48-384 — drained (maxRunners=0)
  c-mt-l-x86iavx512-8-16 — drained (maxRunners=0)
  c-mt-l-x86iavx512-8-64 — drained (maxRunners=0)
  c-mt-l-x86iavx512-94-192 — drained (maxRunners=0)
  c-mt-l-x86iavx512-94-768 — drained (maxRunners=0)
  c-mt-rel-l-arm64g4-16-62 — drained (maxRunners=0)
  c-mt-rel-l-x86iavx512-8-64 — drained (maxRunners=0)

Applying refresh taint to runner nodes...
Updating kubeconfig for pytorch-arc-staging (us-west-1)...
Updated context pytorch-arc-staging in /Users/jschmidt/.kube/config
Tainting ARC runner nodes for arc-staging (pytorch-arc-staging)...
  Taint: deploy.osdc.io/refresh-pending=true:NoSchedule

node/ip-10-0-121-14.us-west-1.compute.internal tainted
  ip-10-0-121-14.us-west-1.compute.internal — tainted

Summary: 1 tainted, 0 already tainted, 0 failed, 1 total runner nodes.

Waiting for in-flight runner pods to drain (timeout: 3600s)...

All runner pods drained (elapsed: 2s).

Summary: 42 AutoscalingRunnerSets drained, 1 nodes tainted, all drained.
```
```
 $  just resume-runners arc-staging
Updating kubeconfig for pytorch-arc-staging (us-west-1)...
Updated context pytorch-arc-staging in /Users/jschmidt/.kube/config
Resuming ARC runner scale sets for arc-staging (pytorch-arc-staging)...

  c-mt-l-arm64g2-6-32 — restored maxRunners=2147483647
  c-mt-l-arm64g3-16-62 — restored maxRunners=2147483647
  c-mt-l-arm64g3-61-463 — restored maxRunners=2147483647
  c-mt-l-arm64g4-16-62 — restored maxRunners=2147483647
  c-mt-l-barm64g4-62-226 — restored maxRunners=2147483647
  c-mt-l-bx86iamx-92-167 — restored maxRunners=2147483647
  c-mt-l-bx86iavx512-88-1000-a100-8 — restored maxRunners=2147483647
  c-mt-l-bx86iavx512-94-344-t4-8 — restored maxRunners=2147483647
  c-mt-l-x86aavx2-189-704-a10g-8 — restored maxRunners=2147483647
  c-mt-l-x86aavx2-29-113-a10g — restored maxRunners=2147483647
  c-mt-l-x86aavx2-29-113-l4 — restored maxRunners=2147483647
  c-mt-l-x86aavx2-45-167-a10g-4 — restored maxRunners=2147483647
  c-mt-l-x86aavx2-45-172-l4-4 — restored maxRunners=2147483647
  c-mt-l-x86aavx512-125-463 — restored maxRunners=2147483647
  c-mt-l-x86iamx-14-27 — restored maxRunners=2147483647
  c-mt-l-x86iamx-16-128 — restored maxRunners=2147483647
  c-mt-l-x86iamx-22-41 — restored maxRunners=2147483647
  c-mt-l-x86iamx-32-128 — restored maxRunners=2147483647
  c-mt-l-x86iamx-46-84 — restored maxRunners=2147483647
  c-mt-l-x86iamx-8-16 — restored maxRunners=2147483647
  c-mt-l-x86iamx-8-32 — restored maxRunners=2147483647
  c-mt-l-x86iamx-8-64 — restored maxRunners=2147483647
  c-mt-l-x86iavx2-40-160 — restored maxRunners=2147483647
  c-mt-l-x86iavx2-8-32 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-11-125-a100 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-16-128 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-16-32 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-2-4 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-22-250-a100-2 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-29-115-t4 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-32-256 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-37-68 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-44-500-a100-4 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-45-172-t4-4 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-46-85 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-48-384 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-8-16 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-8-64 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-94-192 — restored maxRunners=2147483647
  c-mt-l-x86iavx512-94-768 — restored maxRunners=2147483647
  c-mt-rel-l-arm64g4-16-62 — restored maxRunners=2147483647
  c-mt-rel-l-x86iavx512-8-64 — restored maxRunners=2147483647

Removing refresh taint from runner nodes...
Updating kubeconfig for pytorch-arc-staging (us-west-1)...
Updated context pytorch-arc-staging in /Users/jschmidt/.kube/config
Removing refresh taint from ARC runner nodes for arc-staging (pytorch-arc-staging)...
  Taint: deploy.osdc.io/refresh-pending:NoSchedule

node/ip-10-0-121-14.us-west-1.compute.internal untainted
  ip-10-0-121-14.us-west-1.compute.internal — taint removed
  ip-10-0-58-192.us-west-1.compute.internal — not tainted, skipping

Summary: 1 untainted, 1 not tainted, 0 failed, 2 total runner nodes.

Summary: 42 restored, 0 warnings, 0 failures.
```

### CODE REVIEW

File: osdc/justfile [hunk 1]
```diff
@@ -706,6 +706,164 @@ untaint-nodes cluster:
...
+    CLUSTER="{{cluster}}"
+    CNAME=$(uv run {{CFG}} "$CLUSTER" cluster_name)
+
+    # Ensure kubectl is configured for the target cluster
+    just kubeconfig "$CLUSTER"
+
+    echo "Draining ARC runner scale sets for $CLUSTER ($CNAME)..."
+    echo ""
+
+    ARS_NAMES=$(kubectl get autoscalingrunnersets -n arc-runners --no-headers -o custom-columns=NAME:.metadata.name 2>/dev/null || true)
```
Comment (huydhn) [inline osdc/justfile:726]: This include all scaling sets including B200 and H100. Should they be exempted here as they have fixed max capacity?

```
kubectl get autoscalingrunnersets -n arc-runners --no-headers -o custom-columns=NAME:.metadata.name
mt-l-arm64g2-6-32
mt-l-arm64g3-16-62
mt-l-arm64g3-61-463
mt-l-arm64g4-16-62
mt-l-barm64g4-62-226
mt-l-bx86iamx-176-1800-b200-8
mt-l-bx86iamx-176-1800-h100-8
mt-l-bx86iamx-92-167
mt-l-bx86iavx512-88-1000-a100-8
mt-l-bx86iavx512-94-344-t4-8
mt-l-x86aavx2-189-704-a10g-8
mt-l-x86aavx2-29-113-a10g
mt-l-x86aavx2-29-113-l4
mt-l-x86aavx2-45-167-a10g-4
mt-l-x86aavx2-45-172-l4-4
mt-l-x86aavx512-125-463
mt-l-x86iamx-14-27
mt-l-x86iamx-16-128
mt-l-x86iamx-22-225-b200
mt-l-x86iamx-22-225-h100
mt-l-x86iamx-22-41
mt-l-x86iamx-32-128
mt-l-x86iamx-44-450-b200-2
mt-l-x86iamx-44-450-h100-2
mt-l-x86iamx-46-84
mt-l-x86iamx-8-16
mt-l-x86iamx-8-32
mt-l-x86iamx-8-64
mt-l-x86iamx-88-900-b200-4
mt-l-x86iamx-88-900-h100-4
mt-l-x86iavx2-40-160
mt-l-x86iavx2-8-32
mt-l-x86iavx512-11-125-a100
mt-l-x86iavx512-16-128
mt-l-x86iavx512-16-32
mt-l-x86iavx512-2-4
mt-l-x86iavx512-22-250-a100-2
mt-l-x86iavx512-29-115-t4
mt-l-x86iavx512-32-256
mt-l-x86iavx512-37-68
mt-l-x86iavx512-44-500-a100-4
mt-l-x86iavx512-45-172-t4-4
mt-l-x86iavx512-46-85
mt-l-x86iavx512-48-384
mt-l-x86iavx512-8-16
mt-l-x86iavx512-8-64
mt-l-x86iavx512-94-192
mt-l-x86iavx512-94-768
mt-rel-l-arm64g4-16-62
mt-rel-l-x86iavx512-8-64
```

Comment (huydhn) [reply osdc/justfile:726]: Never mind, it seems useful to be able to drain them too

File: osdc/justfile [hunk 2]
```diff
@@ -706,6 +706,164 @@ untaint-nodes cluster:
         echo "Warning: $FAILED node(s) failed to untaint. Re-run to retry."
     fi
 
+# Drain ARC runner scale sets cluster-wide and wait for in-flight runner pods to finish
+drain-runners cluster:
```
Comment (huydhn) [inline osdc/justfile:710]: This would cause queuing to start right away, should we have a `n/Y` confirmation to avoid mistake here?

**Review by huydhn — APPROVED** [review]
LGTM!

File: osdc/justfile (hunk 1 above)
Comment (jeanschmidt) [reply osdc/justfile:726]: yes all of them :)

File: osdc/justfile (hunk 2 above)
Comment (jeanschmidt) [reply osdc/justfile:710]: no, I guess this feels very self explanatory.

if someone runs a command `just drain-runners` I would expect them to be aware that this would cause queueing :)

File: osdc/justfile (hunk 1 above)
Comment (jeanschmidt) [reply osdc/justfile:726]: but when it returns to old max runners, it should read from runner definitions, and as usual: no definition == no limit

File: osdc/justfile (hunk 2 above)
Comment (huydhn) [reply osdc/justfile:710]: True, if you assume a human run it, I was thinking more about me telling my Claude to do something, and it went rouge. We can update https://github.com/pytorch/ci-infra/blob/main/osdc/CLAUDE.md to tell it not to touch this command I guess

**Comment by malfet** [comment]
@claude review this and spot any code duplication
